### ISSUE TITLE
ci instability: GCP upgrades: connection refused on submitting/querying metrics, then missing controller pod on teardown

### ISSUE THREAD

**Issue by nickbp** [opstrace/opstrace#1217]
https://buildkite.com/opstrace/scheduled-main-builds/builds/2809#f2e7d15f-8af2-47e9-83dd-9d2feeed782d

If we check the `bk-artifacts/kubectl_describe_all-pr-upgr-bk-2809-22e-g.log` artifact there arent any controller pods running. There's a `opstrace-controller-9c99c65d4` replicaset that was scaled to 0 but there's no replacement replicaset, so the controller pod wouldn't have been running. Maybe the controller's teardown flag was set wrong in some way?

**Comment by nickbp** [comment]
Another instance, also in GCP upgrades test on a main scheduled build: https://buildkite.com/opstrace/scheduled-main-builds/builds/2819#93952e34-3418-4bf6-8b5b-577c39519a76
```
OldReplicaSets:  <none>
NewReplicaSet:   opstrace-controller-7f6854fd56 (0/0 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  14m   deployment-controller  Scaled up replica set opstrace-controller-7f6854fd56 to 1
  Normal  ScalingReplicaSet  6m8s  deployment-controller  Scaled down replica set opstrace-controller-7f6854fd56 to 0
```
It looks like something fails to relaunch the controller and that's why none of the statefulsets/pvcs/deployments/etc are being deleted, or maybe it's something else entirely.

**Comment by nickbp** [comment]
Actually it looks like in both cases the tests themselves are also very sad. So maybe this is a side effect of the failing tests.

For example in the second build above, test_alerts fails because a metric query returned `connection refused`:
```

[2021-08-08T20:43:18Z] 2021-08-08T20:43:18.127Z [32minfo[39m: Waiting for cortex E2E alerting metric for tenant=default job=testalerts-default-Fh2pJ-job with nonzero value
[2021-08-08T20:43:18Z] 2021-08-08T20:43:18.127Z [32minfo[39m: Cortex query parameter (object):
[2021-08-08T20:43:18Z] {
[2021-08-08T20:43:18Z]   "query": "e2ealerting_webhook_receiver_evaluations_total{job=\"testalerts-default-Fh2pJ-job\"}"
[2021-08-08T20:43:18Z] }
[2021-08-08T20:43:18Z] 2021-08-08T20:43:18.127Z [32minfo[39m: Waiting for query to return a match, deadline in 90s
[2021-08-08T20:43:19Z] 2021-08-08T20:43:19.362Z [32minfo[39m: HTTP resp to GET(https://cortex.default.pr-upgr-bk-2819-83d-g.opstrace.io/api/v1/query?query=e2ealerting_webhook_receiver_evaluations_total%7Bjob%3D%22testalerts-default-Fh2pJ-job%22%7D):
[2021-08-08T20:43:19Z]   status: 502
[2021-08-08T20:43:19Z]   body[:500]: dial tcp 10.224.12.251:80: connect: connection refused
[2021-08-08T20:43:19Z]   headers: {"date":"Sun, 08 Aug 2021 20:43:19 GMT","content-type":"text/plain; charset=utf-8","content-length":"54","connection":"close","strict-transport-security":"max-age=15724800; includeSubDomains"}
[2021-08-08T20:43:19Z]   totalTime: 1.233 s
[2021-08-08T20:43:19Z]   dnsDone->TCPconnectDone: 0.033 s
[2021-08-08T20:43:19Z]   connectDone->reqSent 0 s
[2021-08-08T20:43:19Z]   reqSent->firstResponseByte: 1.064 s
```

The following DD test then fails on both POSTing data and on querying the result with `connection refused` for both:
```

[2021-08-08T20:43:20Z] 2021-08-08T20:43:20.884Z [32minfo[39m: HTTP resp to POST(https://dd.default.pr-upgr-bk-2819-83d-g.opstrace.io/api/v1/series?api_key=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjJmNTgzNDEyZjhkMTY3ZDBiMjdlYTQyNTcxMzY2YjVjZDg2ODFiZmYifQ.eyJpYXQiOjE2Mjg0NTQzNTIsImV4cCI6MTk0NDAzMDM1MiwiYXVkIjoib3BzdHJhY2UtY2x1c3Rlci1wci11cGdyLWJrLTI4MTktODNkLWciLCJpc3MiOiJvcHN0cmFjZS1jbGkiLCJzdWIiOiJ0ZW5hbnQtZGVmYXVsdCJ9.mceFcxr_h8p5z3S6FkGL4LSlqZbnZSHi3PJx-ppV-_X1JCPN-yZrZFFfLJZdlj5mSo9GHvliNY0WII2axhUS-zU0qur3ZxIDYHn_aQs5wQILZ59kbk9ax8HmVJ-4Ze17AoF9jjbaPPD5i5hUEr43N4yFEtO-yH_dhY382jS6wJsLxM-KPIQo8M8JobDUyF7XTQHHeaR-A7iaYQ5C712L5YrPJ0IemwWPTucuqbEFLn5heuSHOiXNqsi4Cu-SY3FF8YK5N1jbY4HzQ3ZU7ksr82xzn9yI1kkxLi-u5CpeyJ09wUZxijc6S6UShDGNRkZLeOE8eL9hitY0OTrM4rVuVw):
[2021-08-08T20:43:20Z]   status: 500
[2021-08-08T20:43:20Z]   body[:500]: error while interacting with remote_write endpoint: Post "http://distributor.cortex.svc.cluster.local/api/v1/push": dial tcp 10.224.0.64:80: connect: connection refused
[2021-08-08T20:43:20Z] 
[2021-08-08T20:43:20Z]   headers: {"date":"Sun, 08 Aug 2021 20:43:20 GMT","content-type":"text/plain; charset=utf-8","content-length":"169","connection":"close","x-content-type-options":"nosniff","strict-transport-security":"max-age=15724800; includeSubDomains"}
[2021-08-08T20:43:20Z]   totalTime: 1.236 s
[2021-08-08T20:43:20Z]   dnsDone->TCPconnectDone: 0.037 s
[2021-08-08T20:43:20Z]   connectDone->reqSent 0 s
[2021-08-08T20:43:20Z]   reqSent->firstResponseByte: 1.088 s
[2021-08-08T20:43:20Z]   
[2021-08-08T20:43:20Z] 2021-08-08T20:43:20.885Z [32minfo[39m: Cortex query parameter (object):
[2021-08-08T20:43:20Z] {
[2021-08-08T20:43:20Z]   "end": "1628455999",
[2021-08-08T20:43:20Z]   "query": "opstrace_dd_test_remote_5jJ5N{job=\"ddagent\"}",
[2021-08-08T20:43:20Z]   "start": "1628452699",
[2021-08-08T20:43:20Z]   "step": "60s"
[2021-08-08T20:43:20Z] }
[2021-08-08T20:43:20Z] 2021-08-08T20:43:20.885Z [32minfo[39m: Waiting for query to return a match, deadline in 90s
[2021-08-08T20:43:22Z] 2021-08-08T20:43:22.014Z [32minfo[39m: HTTP resp to GET(https://cortex.default.pr-upgr-bk-2819-83d-g.opstrace.io/api/v1/query_range?query=opstrace_dd_test_remote_5jJ5N%7Bjob%3D%22ddagent%22%7D&start=1628452699&end=1628455999&step=60s):
[2021-08-08T20:43:22Z]   status: 502
[2021-08-08T20:43:22Z]   body[:500]: dial tcp 10.224.12.251:80: connect: connection refused
[2021-08-08T20:43:22Z]   headers: {"date":"Sun, 08 Aug 2021 20:43:22 GMT","content-type":"text/plain; charset=utf-8","content-length":"54","connection":"close","strict-transport-security":"max-age=15724800; includeSubDomains"}
[2021-08-08T20:43:22Z]   totalTime: 1.128 s
[2021-08-08T20:43:22Z]   dnsDone->TCPconnectDone: 0.033 s
[2021-08-08T20:43:22Z]   connectDone->reqSent 0 s
[2021-08-08T20:43:22Z]   reqSent->firstResponseByte: 1.057 s
```

etc... The end result is a bunch of test failures across suites:
```
[2021-08-08T20:47:35Z] [92m [0m[32m 13 passing[0m[90m (5m)[0m
[2021-08-08T20:47:35Z] [36m [0m[36m 4 pending[0m
[2021-08-08T20:47:35Z] [31m  9 failing[0m
[2021-08-08T20:47:35Z] 
[2021-08-08T20:47:35Z] [0m  1) End-to-end alert tests
[2021-08-08T20:47:35Z]        End-to-end alerts for default and system tenants:
[2021-08-08T20:47:35Z] [0m[31m     SyntaxError: Unexpected token d in JSON at position 0[0m[90m
[2021-08-08T20:47:35Z]       at JSON.parse (<anonymous>)
[2021-08-08T20:47:35Z]       at Object.queryJSONAPI (testutils/index.ts:870:17)
[2021-08-08T20:47:35Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-08T20:47:35Z]       at Object.waitForQueryResult (testutils/index.ts:901:14)
[2021-08-08T20:47:35Z]       at getE2EAlertCountMetric (test_alerts.ts:301:23)
[2021-08-08T20:47:35Z]       at waitForE2EAlertFiring (test_alerts.ts:376:19)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_alerts.ts:444:5)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  2) DD API test suite
[2021-08-08T20:47:35Z]        dd_api_insert_single_ts_fragment:
[2021-08-08T20:47:35Z] [0m[31m     SyntaxError: Unexpected token d in JSON at position 0[0m[90m
[2021-08-08T20:47:35Z]       at JSON.parse (<anonymous>)
[2021-08-08T20:47:35Z]       at Object.queryJSONAPI (testutils/index.ts:870:17)
[2021-08-08T20:47:35Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-08T20:47:35Z]       at Object.waitForQueryResult (testutils/index.ts:901:14)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_dd.ts:297:25)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  3) DD API test suite
[2021-08-08T20:47:35Z]        dd_api_run_agent_container_query_sysuptime:
[2021-08-08T20:47:35Z] [0m[31m     Error: DD agent container setup failed: deadline hit waiting for log needle[0m[90m
[2021-08-08T20:47:35Z]       at startDDagentContainer (test_dd.ts:216:13)
[2021-08-08T20:47:35Z]       at runMicrotasks (<anonymous>)
[2021-08-08T20:47:35Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_dd.ts:326:32)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  4) Prometheus remote_write (push to opstrace cluster) tests
[2021-08-08T20:47:35Z]        local Prom remote_writes metrics with unique label:
[2021-08-08T20:47:35Z] [0m[31m     Error: deadline hit while waiting for expected data on http://127.0.0.1:41541/metrics[0m[90m
[2021-08-08T20:47:35Z]       at Object.sendMetricsWithPromContainer (testutils/index.ts:609:13)
[2021-08-08T20:47:35Z]       at runMicrotasks (<anonymous>)
[2021-08-08T20:47:35Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_prom_remote_write.ts:48:5)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  5) Prometheus remote_write (push to opstrace cluster) tests
[2021-08-08T20:47:35Z]        dummyseries short write, smoke read:
[2021-08-08T20:47:35Z] [0m[31m     Error: Failed to POST PushMessage(md5=d976ca1de7acb95d6433262635357ffc) after 3 attempts[0m[90m
[2021-08-08T20:47:35Z]       at MetricSeriesFragmentPushMessage.postWithRetryOrError (looker/src/metrics/index.ts:223:11)
[2021-08-08T20:47:35Z]       at MetricSeries.postFragmentsToCortex (looker/src/metrics/dummyseries.ts:483:7)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_prom_remote_write.ts:100:5)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  6) Prometheus remote_write (push to opstrace cluster) tests
[2021-08-08T20:47:35Z]        dummyseries short write, fetchAndValidate:
[2021-08-08T20:47:35Z] [0m[31m     Error: Failed to POST PushMessage(md5=8c57b5b99fbf0ffb9fcef45397885847) after 3 attempts[0m[90m
[2021-08-08T20:47:35Z]       at MetricSeriesFragmentPushMessage.postWithRetryOrError (looker/src/metrics/index.ts:223:11)
[2021-08-08T20:47:35Z]       at MetricSeries.postFragmentsToCortex (looker/src/metrics/dummyseries.ts:483:7)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_prom_remote_write.ts:149:5)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  7) Prometheus remote_write (push to opstrace cluster) tests
[2021-08-08T20:47:35Z]        dummyseries long write 1, fetchAndValidate:
[2021-08-08T20:47:35Z] [0m[31m     Error: Failed to POST PushMessage(md5=a4a5a3cd4be6dad572118f987de4a07a) after 3 attempts[0m[90m
[2021-08-08T20:47:35Z]       at MetricSeriesFragmentPushMessage.postWithRetryOrError (looker/src/metrics/index.ts:223:11)
[2021-08-08T20:47:35Z]       at MetricSeries.postFragmentsToCortex (looker/src/metrics/dummyseries.ts:483:7)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_prom_remote_write.ts:173:5)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  8) Prometheus remote_write (push to opstrace cluster) tests
[2021-08-08T20:47:35Z]        dummyseries long write 2, fetchAndValidate:
[2021-08-08T20:47:35Z] [0m[31m     Error: Failed to POST PushMessage(md5=d940632b191b9011ae9549e9e4c97737) after 3 attempts[0m[90m
[2021-08-08T20:47:35Z]       at MetricSeriesFragmentPushMessage.postWithRetryOrError (looker/src/metrics/index.ts:223:11)
[2021-08-08T20:47:35Z]       at MetricSeries.postFragmentsToCortex (looker/src/metrics/dummyseries.ts:483:7)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_prom_remote_write.ts:197:5)
[2021-08-08T20:47:35Z] [0m
[2021-08-08T20:47:35Z] [0m  9) system metrics test suite
[2021-08-08T20:47:35Z]        check if container metrics exist:
[2021-08-08T20:47:35Z] [0m[31m     SyntaxError: Unexpected token d in JSON at position 0[0m[90m
[2021-08-08T20:47:35Z]       at JSON.parse (<anonymous>)
[2021-08-08T20:47:35Z]       at Object.queryJSONAPI (testutils/index.ts:870:17)
[2021-08-08T20:47:35Z]       at runMicrotasks (<anonymous>)
[2021-08-08T20:47:35Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-08T20:47:35Z]       at Object.waitForQueryResult (testutils/index.ts:901:14)
[2021-08-08T20:47:35Z]       at Context.<anonymous> (test_systemmetrics.ts:51:25)
```

**Comment by nickbp** [comment]
The cortex logs in `clusterlogs_cortex-pr-upgr-bk-2819-83d-g.log` are also full of `connection refused` errors against these IPs (cross-referenced to their pods):
- 10.220.1.31: not mentioned in describe list
- 10.220.1.36: ruler-586db6f4c8-v526x, had two restarts (last at 20:35:44) but was Running from 20:35:57
- 10.220.2.23: not mentioned in describe list
- 10.220.2.25: ruler-586db6f4c8-pnd69, had two restarts (last at 20:35:44, same as above) but was Running from 20:35:58

So those logs in cortex might not be definitive issues and could have just been part of the system coming up.

**Comment by nickbp** [comment]
Meanwhile if we check the test errors from earlier, they mention `10.224.12.251` which maps to this Service with no Endpoints:
```
Name:              query-frontend
Namespace:         cortex
Labels:            job=cortex.query-frontend
                   name=query-frontend
Annotations:       cloud.google.com/neg: {"ingress":true}
                   opstrace: no-update
Selector:          name=query-frontend
Type:              ClusterIP
IP Families:       <none>
IP:                10.224.12.251
IPs:               <none>
Port:              http  80/TCP
TargetPort:        80/TCP
Endpoints:         <none>
Port:              grpc  9095/TCP
TargetPort:        9095/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>
```

The equivalent `cortex/querier` service also has no endpoints. Indeed, there are no `cortex/query-frontend` pods, nor `cortex/querier` pods. Maybe something caused these to not be deployed? Is this expected?

**Comment by nickbp** [comment]
https://buildkite.com/opstrace/scheduled-main-builds/builds/2821#85a5d558-8043-495f-8fbf-5c1b8845b8ca

**Comment by jgehrcke** [comment]
Big thanks for the debugging effort so far. Looks like this really needs to be understood, with some priority.

**Comment by sreis** [comment]
The `create` command deploys the controller with this image tag:
```
[2021-08-07T04:24:06Z]   "controller_image": "opstrace/controller:22e9a821-ci",
``` 

Then the `upgrade` command needs to scale down the opstrace controller to be able to upgrade the cortex deployment to use the cortex-operator.

```
[2021-08-07T04:48:00Z] 2021-08-07T04:48:00.026Z [32minfo[39m: controller deployment: Available: True (Deployment has minimum availability.)
[2021-08-07T04:48:00Z] 2021-08-07T04:48:00.026Z [32minfo[39m: controller deployment: Progressing: True (ReplicaSet "opstrace-controller-9c99c65d4" has successfully progressed.)
[2021-08-07T04:48:00Z] 2021-08-07T04:48:00.027Z [32minfo[39m: controller deployment: replicas: undefined, of which are READY: 0
[2021-08-07T04:48:00Z] 2021-08-07T04:48:00.027Z [32minfo[39m: desired state reached, continue
```

And then it skips updating the deployment because it's the same image tag and never restarts the controller.

```
[2021-08-07T04:48:00Z] 2021-08-07T04:48:00.825Z [32minfo[39m: controller image is already at desired version: opstrace/controller:22e9a821-ci
```

This uncovered a few things. Looks like the `upgrade` command is not handling the case where the opstrace-controller image was not updated (upgrade to the same version?). This means there's an issue in the Buildkite or build pipeline that uncovered a bug in the `upgrade` command logic because the base cluster version should not match the version in the main branch. Another issue is that the tests should have been skipped since the upgrade failed.

I'll keep digging.

**Comment by sreis** [comment]
Looks like #1221 explains why the upgrade test was trying to upgrade to the same version of the controller.

**Comment by jgehrcke** [comment]
Similar symptom: https://github.com/opstrace/opstrace/issues/1222

**Comment by jgehrcke** [comment]
> Looks like #1221 explains why the upgrade test was trying to upgrade to the same version of the controller.

Huh! We run the "PR" pipeline also as the scheduled build from main. And need to carefully distinguish:

1) when we run the PR pipeline on a PR then FROM and TO should _always_ have different version strings, even when FROM is  `https://go.otr.dev/cli-latest-linux` (which is the latest build from `main`).

2) when we run the PR pipeline as a scheduled build on `main` then of course it can happen that the FROM and TO version strings match. This should exit the upgrade command early anyway, right?

So. This ticket here so far only shows failures from scheduled builds from main, i.e. scenario (2) was possible.

But in #1222 I just found a similar symptom, and that happened on a PR.

**Comment by sreis** [comment]
Closing since we haven't seen this after #1223 was merged.
